### PR TITLE
Use JDK LdapName and Rdn instead of our classes

### LINKED ISSUE #21797: DN with RDN that contains trailing backslash is imported incorrectly into Keycloak
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

ldap

### Describe the bug

When connecting Keycloak to an AD LDAP server with a DN that has a trailing backslash in one of the RDNs, the DN is imported incorrectly. 

AD DN: `CN=Test User\\\\,OU=Users,DC=example,DC=com`

But when imported and viewed in Keycloak it is no longer a valid DN. 

Keycloak LDAP_ENTRY_DN: `CN=Test User\\\\,OU,DC=example,DC=com`

This causes login and parsing issues for the entity that is imported. 

### Version

20.0.2, 21.1.2

### Expected behavior

Expected behavior is that the entity with a DN containing an RDN with the trailing backslash would be imported correctly with the same DN as seen in the AD server. eg `CN=Test User\\\\,OU=Users,DC=example,DC=com`

### Actual behavior

Actual behavior is that the entity with a DN containing an RDN with the trailing backslash is imported incorrectly and the entity cannot be used in Keycloak. eg `CN=Test User\\\\,OU,DC=example,DC=com`

### How to Reproduce?

Install Keycloak and connect to an Active Directory LDAP server that has an entity with a DN containing an RDN with a trailing backslash. eg `CN=Test User\\\\,OU=Users,DC=example,DC=com`

### Anything else?

_No response_

### LINKED ISSUE #21818: DN with RDN that contains trailing space is imported incorrectly into Keycloak
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

ldap

### Describe the bug

When connecting Keycloak to an AD LDAP server with a DN that has a trailing space in one of the RDNs, the DN is imported incorrectly.

AD DN: 'CN=Test User\\\\ ,OU=Users,DC=example,DC=com'

But when imported and viewed in Keycloak the space has been trimmed.

Keycloak LDAP_ENTRY_DN: 'CN=Test User\\\\,OU=Users,DC=example,DC=com'

This causes login issues for the entity that is imported.

### Version

20.0.2, 21.1.2

### Expected behavior

Expected behavior is that the entity with a DN containing an RDN with the trailing space would be imported correctly with the same DN as seen in the AD server. eg 'CN=Test User\\\\ ,OU=Users,DC=example,DC=com'

### Actual behavior

Actual behavior is that the entity with a DN containing an RDN with the trailing space is imported incorrectly and the entity cannot be used in Keycloak. The space at the end of the RDN is trimmed before committing to the database. eg 'CN=Test User\\\\,OU=Users,DC=example,DC=com'

### How to Reproduce?

Install Keycloak and connect to an Active Directory LDAP server that has an entity with a DN containing an RDN with a trailing space. eg 'CN=Test User\\\\ ,OU=Users,DC=example,DC=com'

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by rmartinc**
Closes https://github.com/keycloak/keycloak/issues/21797
Closes https://github.com/keycloak/keycloak/issues/21818

Just a draft to see if we are interested in this (or something similar).

Looking those two issues about LDAP I wondered why we are using `LDAPDn` and `LDAPDn.RDN` instead of `LdapName` and `Rdn` default java classes. I did this quick test and it passed all the CI (https://github.com/rmartinc/keycloak/actions/runs/6380208154). Three commits are submitted with three different levels of replacement:

1.  `LDAPDn` and `LDAPDn.RDN` wrap the java `LdapName/Rdn` classes.
2. API change to use `LdapName` instead of `String` and avoid extra conversions (LdapDn/LdapName -> String -> LdapName).
3. Remove completely `LDAPDn` class. Use LdapName/Rdn instead and I moved the interesting `LDAPDn` methods to `LDAPUtils` class.

I'm changing a lot the API inside the ldap federation (not in point 1, just in 2 and 3). Don't know if we are interested in maintain compatiblity in those classes. I'm happy with 1 and 2 only. I just did 3 to see if it was feasible.

@keycloak/core @keycloak/store WDYT?

### CODE REVIEW

**Review by mposolda — DISMISSED** [review]
This approach is fine to me, so I am approving it.

Now when seeing the PR, I recall that one of the reasons for our own class `LDAPDn` was missing methods like `getFirstRdn`, `addFirstDn` etc on the default Java `LdapName` class. One alternative approach for this limitation might be to keep our class `LDAPDn` and just make it wrapper over Java `LdapName` . But your approach with using `LdapName` directly and have the helper methods on `LDAPUtils` works for me as well. So approving this PR unless store team has some objections. Leaving to them to review this (CC @mhajas @ahus1)

There are also some "Code QL" failing checks, however those should be fine IMO. Unless the "Code QL" check is broken for all other PRs when this one is merged, which would be quite bad, but AFAIK it doesn't work that way as "Code QL" checks only changes in the code of the particular PR?

**Comment by rmartinc** [comment]
@mposolda Yes, that's the difference between commit 2 and 3. If we add the commit 3 the LDAPDn is removed completely and the Utils are used. We can remove that commit and just maintain commits 1 and 2, then the situation is that LDAPDn and RDN are maintained but wrapping an LdapName and Rdn respectively. As I commented I'm happy with commits 1 and 2 only. So we can do whatever you guys prefer.

I think that the Code QL error is produced because the search methods are public and the filter is just a String passed to the method. So I would say that the error was there before the changes in this PR.

**Comment by mposolda** [comment]
@mhajas @ahus1 What is your preference? Keep the `LDAPDn` class as wrapper over java `LdapName` OR use `LdapName` everywhere and methods like `getFirstRdn` etc on `LDAPUtils` class?

**Comment by mhajas** [comment]
Thanks for the PR @rmartinc! I didn't have a chance to go through this PR properly, I can get to that probably next week. In general, I like getting rid of custom code and replacing it with some standard implementation :+1:. 

Do we know what was the reason for introducing `LdapDn` class before? Is it because the standard implementation was not sufficient before and now it is? Or do we need some custom handling?

**Comment by rmartinc** [comment]
> Do we know what was the reason for introducing LdapDn class before? Is it because the standard implementation was not sufficient before and now it is? Or do we need some custom handling?

I personally don't know the answer and that's why I did the 3rd commit, to see if completely removing our classes run the CI OK. I don't know of limitations in those classes since jdk-8. @mposolda commented about the methods in LDAPDn and RDN (getFirstRdn, getParentDn, isDescendantOd,...) but we can maintain those methods but wrapping LdapName/Rdn (commits 1+2 only).

**Comment by mposolda** [comment]
@mhajas Yes, I recall the motivation for introducing `LdapDN` was some missing utilities like `getFirstRdn` and others (which are moved in @rmartinc in class `LDAPUtils`). And maybe some other limitations in escaping in default java class `LdapName` in historic java versions, but those are likely already improved in Java 17. 

So in summary we see options like:
1) Keep our classes like `LDAPDn` and make it a wrapper over java `LDapName` and keep the helper methods (like `getFirstRdn` here)
2) Remove `LDAPDn` and use default java classes and possibly have helper methods on `LDAPUtils` class.

@rmartinc uses approach (1) in first 2 commits of this PR and approach (2) in 3rd commit on top of that. So we can choose whatever you prefer. I am maybe slightly for approach  (1), but not too strong on this...

**Comment by ahus1** [comment]
I'm also slightly preferring (1) like @mposolda. Happy to see fewer usages of an untyped `String` after this change.

**Comment by rmartinc** [comment]
@mposolda I have updated the PR to use the preferred option 1. Please re-run the two windows run that failed because of the node exe issue. Moving out of draft.

**Review by mposolda — DISMISSED** [review]
@rmartinc Looks good, hope the tests passed once they are re-run. Thanks!

**Comment by abstractj** [comment]
@mposolda @rmartinc the CodeQL issues reported were not introduced by the changes here. I'll share with you guys a proposal to see if we can fix it.

**Review by ahus1 — DISMISSED** [review]
Approving from the store teams point of view. The core team should merge it once they think it is ready.

**Review by mposolda — APPROVED** [review]

**Pull request merged**
